File: requel-angular/src/app/shared/validation-limits.ts
/*
 * This file is part of Requel - the Collaborative Requirements
 * Elicitation System.
 *
 * Copyright 2026 Ron Regan Jr. All Rights Reserved.
 *
 * Requel is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Requel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Requel. If not, see <http://www.gnu.org/licenses/>.
 *
 */

/**
 * Client-side validation bounds, each one mirroring a constraint the server actually
 * enforces (issue #132).
 *
 * The rule for this file: **every entry names its backend source.** A client-side cap
 * with no server-side counterpart rejects input the server would have accepted, which
 * is worse than no cap at all — the user is blocked by a rule that does not exist. So
 * a bound only appears here once something in `modules/` enforces it, and the comment
 * says where, so the next person can diff this against the annotations instead of
 * guessing which numbers are real.
 *
 * What is deliberately ABSENT: a max length for artifact `text`. `AbstractTextEntity.getText()`
 * is `@Lob` (`longtext`), so there is no server-side bound to mirror — and a cap here with no
 * server-side counterpart is the exact failure this file exists to prevent. The `name` bound
 * below arrived with #171, which also made these DTO constraints actually run: before it, nothing
 * in the backend invoked a `Validator` at all.
 */

/**
 * Maximum password length.
 *
 * Source: `UserImpl.MAX_PASSWORD_LENGTH` (user-jpa). `UserImpl.isValidPassword` accepts
 * a password that is non-blank and no longer than this, and that is the *only* password
 * rule the server applies — no complexity requirement, no minimum beyond non-blank. The
 * client mirrors exactly that, so a password the user is allowed to set is never
 * rejected by the form first. A real password policy is a product decision, not
 * something to invent here.
 */
export const PASSWORD_MAX_LENGTH = 128;

/**
 * Minimum number of roles a user must hold.
 *
 * Source: `UserImpl:385`, `@Size(min = 1)` on the roles collection — the one size
 * constraint the backend already had before #171.
 */
export const USER_ROLES_MIN = 1;

/**
 * Maximum length of an artifact `name`.
 *
 * Source: `ValidationLimits.ARTIFACT_NAME_MAX` (platform-core), applied as
 * `@Size(max = 255)` to the `name` of every artifact entity — goals, stories, actors, use cases,
 * scenarios, terms, reports, stakeholders, projects, teams, tag categories, organizations, users —
 * and to the matching `Edit*Input` DTO fields. It is not an invented product limit: every one of
 * those columns is already `varchar(255)`, so this is the database's own bound made visible to the
 * user instead of surfacing as a driver error.
 *
 * Also applies to `username`, which shares the 255-character column on `users`.
 */
export const ARTIFACT_NAME_MAX_LENGTH = 255;

File: modules/platform-core/src/main/java/com/rreganjr/validator/ValidationLimits.java
/*
 * This file is part of Requel - the Collaborative Requirements
 * Elicitation System.
 *
 * Copyright 2026 Ron Regan Jr. All Rights Reserved.
 *
 * Requel is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Requel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Requel. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.rreganjr.validator;

/**
 * Shared bounds for bean validation constraints (issue #171).
 *
 * <p>The point of putting these in one place is that the same number has to appear on both sides of
 * every constraint — on the JPA entity, where it is enforced at flush time, and on the input DTO,
 * where {@code CommandInputValidator} enforces it before the command runs — and a third time in the
 * Angular client, whose {@code shared/validation-limits.ts} mirrors it so the form reports an
 * over-long value inline instead of waiting for a round trip. Three copies of a literal 255 across
 * three languages is a number that drifts. One constant is a number the client file can name.
 *
 * <p>These are constraint bounds, not schema. The database column widths they match are owned by
 * Flyway (`modules/requel-app/src/main/resources/db/migration/`), and nothing here changes DDL.
 */
public final class ValidationLimits {

    /**
     * Maximum length of an artifact {@code name}.
     *
     * <p>255 is not chosen, it is <em>observed</em>: every artifact name column in
     * {@code V1__init.sql} is already {@code varchar(255)} — goals, stories, actors, use cases,
     * scenarios, terms, reports, stakeholders, projects, teams, organizations, users — as is
     * {@code tag_category.name} in {@code V13__tagging.sql}. Matching the DDL is what makes this
     * constraint free of a migration: it converts what is currently a driver-level truncation error
     * into a field-level validation message, and changes nothing about what the database accepts.
     *
     * <p>Do not raise this past ~767 without checking indexes.
     * {@code goals}, {@code stories}, {@code actors}, {@code usecases}, {@code scenarios},
     * {@code terms} and {@code stakeholders} all carry a {@code UNIQUE KEY (projectordomain_id,
     * name)}; at {@code varchar(255)} utf8mb4 that is 1020 bytes, comfortably inside InnoDB's
     * 3072-byte limit, but a wider column would not be.
     *
     * <p>Deliberately absent: any bound on artifact {@code text}.
     * {@code AbstractTextEntity.getText()} is {@code @Lob} ({@code longtext}), so a cap here would
     * be invented rather than mirrored — and a client-side cap with no server-side counterpart
     * rejects input the server would have accepted.
     */
    public static final int ARTIFACT_NAME_MAX = 255;

    /**
     * Message template for a length violation. {@code {max}} is interpolated by the validator, so
     * this stays correct if {@link #ARTIFACT_NAME_MAX} changes.
     *
     * <p>Worded to match the hand-written entity messages already in the codebase ("a unique name is
     * required.", "one or more roles must be selected.") rather than the bean-validation default
     * ("size must be between 0 and 255"), because these strings are rendered under the field in the
     * UI, not read in a log.
     */
    public static final String LENGTH_MESSAGE = "must be {max} characters or fewer.";

    private ValidationLimits() {
        // constants only
    }
}
